kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,363 @@
+            "hosts": {},
+            "vars": {
+                "ansible_ssh_common_args": "-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no"
+            },
+        },
+        "kube-node": {"hosts": {}},
+        "etcd": {"hosts": {}},
+        "vault": {"hosts": {}},
+        "k8s-cluster": {"children": {"kube-node": None,
+                                     "kube-master": None}},
+    }
+    for master in outputs["masters"]:
+        conf["all"]["hosts"][master["hostname"]] = {
+            "access_ip": master["ip"],
+            "ansible_host": master["ip"],
+            "ansible_user": "ubuntu",
+            "ansible_become": True,
+        }
+        conf["kube-master"]["hosts"][master["hostname"]] = None
+        conf["etcd"]["hosts"][master["hostname"]] = None
+        conf["vault"]["hosts"][master["hostname"]] = None
+    for slave in outputs["slaves"]:
+        conf["all"]["hosts"][slave["hostname"]] = {
+            "ansible_host": slave["ip"],
+            "ansible_user": "ubuntu",
+            "ansible_become": True,
+        }
+        conf["kube-node"]["hosts"][slave["hostname"]] = None
+    ssh_args_fmt = ("-o ProxyCommand=\"ssh ubuntu@{} "
+                    "-o UserKnownHostsFile=/dev/null "
+                    "-o StrictHostKeyChecking=no -W %h:%p\" "
+                    "-o UserKnownHostsFile=/dev/null "
+                    "-o StrictHostKeyChecking=no")
+    ssh_args = ssh_args_fmt.format(outputs["masters"][0]["ip"])
+    conf["kube-node"]["vars"] = {"ansible_ssh_common_args": ssh_args}
+    return conf
+
+
+class Heat:
+    """Openstack Heat client wrapper."""
+
+    def __init__(self, **kwargs):
+        self.connection = openstack.connection.Connection(**kwargs)
+
+    def _get_template_and_files(self, package_name):
+        # NOTE(sskripnick)
+        # we need to custom load and decode all files because of this bug:
+        # https://storyboard.openstack.org/#!/story/2002002
+        package = importlib.import_module(package_name)
+        path = list(package.__path__)[0]
+        files = {}
+        for root, dirs, _files in os.walk(path):
+            for f in _files:
+                fpath = os.path.join(root, f)
+                files["file://" + fpath] = open(fpath, "rb").read().decode("utf8")
+        return os.path.join(path, "template.yaml"), files
+
+    def create(self, *, stack_name, package_name, **kwargs):
+        template_file, files = self._get_template_and_files(package_name)
+        try:
+            return self.connection.create_stack(
+                name=stack_name,
+                template_file=template_file,
+                files=files,
+                **kwargs
+            )
+        except openstack.exceptions.BadRequestException as e:
+            # NOTE(sskripnick)
+            # handle this exception separately because useful information
+            # may be found only in e.details
+            logger.exception("Failed to create stack: %s" % e.details)
```

Comment:
format()

Reply:
https://www.python.org/dev/peps/pep-3101/  tells us:  "intended as a replacement for the existing '%' string"  
and  % is called old in the python3 docs

Reply:
it is about to follow one style for a particular project, for kqueen format is used for formatting strings

Reply:
@katyafervent 

> for kqueen format is used for formatting strings

Is this mentioned somewhere in the docs? Is there an automatic check for this?